sale_coupon_auto_refresh/models/sale_order.py

```diff
@@ -9,22 +11,55 @@ class SaleOrder(models.Model):
     # Used in UI to hide the manual button
     auto_refresh_coupon = fields.Boolean(related="company_id.auto_refresh_coupon",)
 
-    def _auto_refresh_coupons(self):
-        self.filtered(
-            lambda x: x.state in ("draft", "sent") and x.auto_refresh_coupon
-        ).recompute_coupon_lines()
-
     @api.model_create_multi
     def create(self, vals_list):
-        """Create or refresh coupon lines on create"""
         orders = super().create(vals_list)
-        orders.with_context(skip_auto_refresh_coupons=True)._auto_refresh_coupons()
+        orders._auto_refresh_coupons()
         return orders
 
     def write(self, vals):
-        """Create or refresh coupon lines after saving"""
         res = super().write(vals)
-        if self.env.context.get("skip_auto_refresh_coupons"):
-            return res
-        self.with_context(skip_auto_refresh_coupons=True)._auto_refresh_coupons()
+        self._auto_refresh_coupons()
+        return res
+
+    def _auto_refresh_coupons(self):
+        if not self.env.context.get("skip_auto_refresh_coupons"):
+            orders = self.filtered(type(self)._allow_recompute_coupon_lines)
```

Comment:
I don't get this weird construction. Why not?

```suggestion
            orders = self.filtered(lambda x: x._allow_recompute_coupon_lines())
```

Reply:
@pedrobaeza it's actually the same thing: using `type(self).func` allows passing the function `func` (as an object retrieved from the class `type(self)`) without instantiating a lambda function (which is better, performance and memory wise).
BTW please note, as I wrote above, that using `self.func` will not achieve the same result (it will raise an error).

Reply:
OK, let's continue this way.